src/include/concurrency/centralized_epoch_manager.h

```diff
@@ -0,0 +1,214 @@
+//===----------------------------------------------------------------------===//
+//
+//                         Peloton
+//
+// centralized_epoch_manager.h
+//
+// Identification: src/include/concurrency/centralized_epoch_manager.h
+//
+// Copyright (c) 2015-16, Carnegie Mellon University Database Group
+//
+//===----------------------------------------------------------------------===//
+
+
+#pragma once
+
+#include <thread>
+#include <vector>
+
+#include "common/macros.h"
+#include "type/types.h"
+#include "common/logger.h"
+#include "common/platform.h"
+#include "common/init.h"
+#include "common/thread_pool.h"
+#include "concurrency/epoch_manager.h"
+
+namespace peloton {
+namespace concurrency {
+
+/*
+Epoch queue layout:
+ current epoch               queue tail                reclaim tail
+/                           /                          /
++--------+--------+--------+--------+--------+--------+--------+-------
+| head   | safety |  ....  |readonly| safety |  ....  |gc usage|  ....
++--------+--------+--------+--------+--------+--------+--------+-------
+New                                                   Old
+
+Note:
+1) Queue tail epoch and epochs which is older than it have 0 rw txn ref count
+2) Reclaim tail epoch and epochs which is older than it have 0 ro txn ref count
+3) Reclaim tail is at least 2 turns older than the queue tail epoch
+4) Queue tail is at least 2 turns older than the head epoch
+*/
+
+class CentralizedEpochManager : public EpochManager {
+  CentralizedEpochManager(const CentralizedEpochManager&) = delete;
+  static const int safety_interval_ = 2;
+
+struct Epoch {
+  std::atomic<int> ro_txn_ref_count_;
+  std::atomic<int> rw_txn_ref_count_;
+  cid_t max_cid_;
+
+  Epoch(): 
+    ro_txn_ref_count_(0), 
+    rw_txn_ref_count_(0),
+    max_cid_(0) {}
+
+  Epoch(const Epoch &epoch): 
+    ro_txn_ref_count_(epoch.ro_txn_ref_count_.load()), 
+    rw_txn_ref_count_(epoch.rw_txn_ref_count_.load()),
+    max_cid_(0) {}
+
+  void Init() {
+    ro_txn_ref_count_ = 0;
+    rw_txn_ref_count_ = 0;
+    max_cid_ = 0;
+  }
+};
+
+public:
+  CentralizedEpochManager()
+    : epoch_queue_(epoch_queue_size_),
+      queue_tail_(0), 
+      reclaim_tail_(0), 
+      current_epoch_(0),
+      queue_tail_token_(true), 
+      reclaim_tail_token_(true),
+      max_cid_ro_(READ_ONLY_START_CID), 
+      max_cid_gc_(0), 
+      is_running_(false) {}
+
+
+  static CentralizedEpochManager &GetInstance() {
+    static CentralizedEpochManager epoch_manager;
+    return epoch_manager;
+  }
+
+  virtual void Reset(const size_t &current_epoch) override {
+    current_epoch_ = current_epoch;
+  }
+
+  virtual void StartEpoch(std::unique_ptr<std::thread> &epoch_thread) override {
+    LOG_TRACE("Starting epoch");
```

Comment:
No. The current implemtation is that each component must acquire (dedicated) thread from thread pool.